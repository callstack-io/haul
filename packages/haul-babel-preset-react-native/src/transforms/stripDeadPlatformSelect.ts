import * as babel from '@babel/core';

export default function stripDeadPlatformSelect() {
  return {
    visitor: {
      CallExpression(
        path: babel.NodePath<babel.types.CallExpression>,
        state: { opts: { platform: string } }
      ) {
        if (
          babel.types.isMemberExpression(path.node.callee) &&
          babel.types.isIdentifier(path.node.callee.object) &&
          babel.types.isIdentifier(path.node.callee.property) &&
          path.node.callee.object.name === 'Platform' &&
          path.node.callee.property.name === 'select' &&
          babel.types.isObjectExpression(path.node.arguments[0])
        ) {
          const platformsSpecs = (path.node
            .arguments[0] as unknown) as babel.types.ObjectExpression;

          let canStripPlatformSelect = true;
          let targetCase: babel.types.ObjectProperty | undefined;
          let defaultCase: babel.types.ObjectProperty | undefined;
          const additionalProperties: babel.types.ObjectExpression['properties'] = [];

          platformsSpecs.properties.forEach(property => {
            if (
              babel.types.isObjectProperty(property) &&
              babel.types.isIdentifier(property.key)
            ) {
              if (property.key.name === 'default') {
                defaultCase = property;
              } else if (property.key.name === state.opts.platform) {
                targetCase = property;
              }
            } else {
              canStripPlatformSelect = false;
              additionalProperties.push(property);
            }
          });

          // If we got exact mach, we can strip the rest
          if (targetCase) {
            canStripPlatformSelect = true;
          }

          if (!targetCase && !defaultCase && canStripPlatformSelect) {
            path.replaceWithSourceString('undefined');
          } else if (canStripPlatformSelect && (targetCase || defaultCase)) {
            path.replaceWith((targetCase || defaultCase)!.value as any);
          } else {
            platformsSpecs.properties = [
              targetCase || defaultCase,
              ...additionalProperties,
            ].filter(Boolean) as babel.types.ObjectExpression['properties'];
          }
        }
      },
    },
  };
}
