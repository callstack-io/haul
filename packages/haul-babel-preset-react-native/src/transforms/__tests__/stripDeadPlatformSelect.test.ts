import stripDeadPlatformSelect from '../stripDeadPlatformSelect';
import { transformSync } from '@babel/core';

const fixture1 = `
const value = Platform.select({
  android: 'android',
  ios: () => {},
  default: { test: true },
  web: 'what?'
});
`;

const fixture2 = `
const defaults = { default: 'default' };
const value = Platform.select({
  ios: () => {},
  ...defaults,
});
`;

const fixture3 = `
const value = Platform.select({
  android: () => {},
});
`;

const fixture4 = `
const value = Platform.select({
  ios: false,
  ['andr' + 'oid']: () => {},
});
`;

const fixture5 = `
const value = Platform.select({
  default: 'default',
  ['andr' + 'oid']() {}
});
`;

const fixture6 = `
const value = Platform.select({
  ios: false,
  android() {}
});
`;

describe('stripDeadPlatformSelect transform', () => {
  it('should replace Platform.select with value for default case if there is no matching platform', () => {
    const { code } = transformSync(fixture1, {
      plugins: [[stripDeadPlatformSelect, { platform: 'custom' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch(/const value = {\s*test: true\s*}/);
    expect(code).not.toMatch('Platform.select');
    expect(code).toMatchSnapshot();
  });

  it('should replace Platform.select with value for matching platform', () => {
    const { code } = transformSync(fixture2, {
      plugins: [[stripDeadPlatformSelect, { platform: 'ios' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch('const value = () => {};');
    expect(code).not.toMatch('Platform.select');
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select', () => {
    const { code } = transformSync(fixture2, {
      plugins: [[stripDeadPlatformSelect, { platform: 'android' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch('...defaults');
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).toMatchSnapshot();
  });

  it('should replace Platform.select with undefined if no cases are matching', () => {
    const { code } = transformSync(fixture3, {
      plugins: [[stripDeadPlatformSelect, { platform: 'ios' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toEqual('const value = undefined;');
  });

  it('should remove non-matching platforms but leave Platform.select with unknown cases 1', () => {
    const { code } = transformSync(fixture4, {
      plugins: [[stripDeadPlatformSelect, { platform: 'android' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch("['andr' + 'oid']: () => {}");
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select with unknown cases and default case', () => {
    const { code } = transformSync(fixture5, {
      plugins: [[stripDeadPlatformSelect, { platform: 'android' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch("['andr' + 'oid']() {}");
    expect(code).toMatch('Platform.select');
    expect(code).toMatch("default: 'default',");
    expect(code).toMatchSnapshot();
  });

  it('should remove non-matching platforms but leave Platform.select with method', () => {
    const { code } = transformSync(fixture6, {
      plugins: [[stripDeadPlatformSelect, { platform: 'android' }]],
      babelrc: false,
      configFile: false,
    }) || { code: '' };

    expect(code).toMatch('android() {}');
    expect(code).toMatch('Platform.select');
    expect(code).not.toMatch('ios');
    expect(code).toMatchSnapshot();
  });
});
