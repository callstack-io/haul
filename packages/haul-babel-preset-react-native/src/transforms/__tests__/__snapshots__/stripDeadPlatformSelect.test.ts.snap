// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripDeadPlatformSelect transform should remove non-matching platforms but leave Platform.select 1`] = `
"const defaults = {
  default: 'default'
};
const value = Platform.select({ ...defaults
});"
`;

exports[`stripDeadPlatformSelect transform should remove non-matching platforms but leave Platform.select with method 1`] = `
"const value = Platform.select({
  android() {}

});"
`;

exports[`stripDeadPlatformSelect transform should remove non-matching platforms but leave Platform.select with unknown cases 1 1`] = `
"const value = Platform.select({
  ['andr' + 'oid']: () => {}
});"
`;

exports[`stripDeadPlatformSelect transform should remove non-matching platforms but leave Platform.select with unknown cases and default case 1`] = `
"const value = Platform.select({
  default: 'default',

  ['andr' + 'oid']() {}

});"
`;

exports[`stripDeadPlatformSelect transform should replace Platform.select with value for default case if there is no matching platform 1`] = `
"const value = {
  test: true
};"
`;

exports[`stripDeadPlatformSelect transform should replace Platform.select with value for matching platform 1`] = `
"const defaults = {
  default: 'default'
};

const value = () => {};"
`;
